src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala

```diff
@@ -127,6 +127,17 @@ class ResourceService(private val resourceTypes: Map[ResourceTypeName, ResourceT
     }
   }
 
+  def addOrRemoveUserFromPolicy(resourceType: ResourceType, policyName: AccessPolicyName, resource: Resource, email: String, userInfo: UserInfo, add: Boolean = true): Future[AccessPolicy] = {
```

Comment:
Loading in the entire policy and overwriting it can lead to problem if multiple requests come in at the same time. They will clobber each other. I understand that it is convenient to do so because we already have the overwrite code. But it is also easy to just add or remove a member from a policy, in fact the [add/remove]GroupMember functions in JndiDirectoryDAO can be used directly.

Be sure to handle errors when a user being added already exists or a user being removed does not exist. I think both these cases should be treated as OK.

Also in each case call the cloudExtension.onGroupUpdate function.

And a note on the existing implementation: try to avoid the use of flags (the add parameter). 2 separate functions are easier to follow and you can do code reuse within. I would be less critical if it were a private function but only slightly less ;)

Last, write some unit tests.